Signed-out callback path coverage (#34385)
dotnet/AspNetCore.Docs: aspnetcore/blazor/security/blazor-web-app-with-entra.md

The callback path (`CallbackPath`) must match the redirect URI (login callback path) configured when registering the application in the Entra or Azure portal. Paths are configured in the **Authentication** blade of the app's registration. The default value of `CallbackPath` is `/signin-oidc` for a registered redirect URI of `https://localhost/signin-oidc` (a port isn't required).

> The <xref:Microsoft.AspNetCore.Builder.OpenIdConnectOptions.SignedOutCallbackPath%2A> (configuration key: "`SignedOutCallbackPath`") is the request path within the app's base path where the user agent is returned after signing out from Entra. The preceding example doesn't set a value for the path because the default value of "`/signout-callback-oidc`" is used.

Reviewer: ```suggestion
The <xref:Microsoft.AspNetCore.Builder.OpenIdConnectOptions.SignedOutCallbackPath%2A> (configuration key: "`SignedOutCallbackPath`") is the request path within the app's base path intercepted by the OpenID Connect handler where the user agent is first returned after signing out from Entra. The preceding example doesn't set a value for the path because the default value of "`/signout-callback-oidc`" is used. After intercepting the `/signout-callback-oidc` the OpenID Connect handler finally redirects to the <xref:Microsoft.AspNetCore.Builder.OpenIdConnectOptions.SignedOutRedirectUri%2A> or <xref:Microsoft.AspNetCore.Authentication.AuthenticationProperties.RedirectUri%2A> if specified.
```

Author: I'll adjust all three spots in one go on my branch for this.

I'll change "OpenID Connect" to "OIDC" because the abbreviation is used throughout the articles.

> The <xref:Microsoft.AspNetCore.Builder.OpenIdConnectOptions.SignedOutCallbackPath%2A> (configuration key: "`SignedOutCallbackPath`") is the request path within the app's base path intercepted by the OpenID Connect handler where the user agent is first returned after signing out from Entra. The sample app doesn't set a value for the path because the default value of "`/signout-callback-oidc`" is used. After intercepting the request, the OpenID Connect handler finally redirects to the <xref:Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutRedirectUri%2A> or <xref:Microsoft.AspNetCore.Authentication.AuthenticationProperties.RedirectUri%2A>, if specified.

Reviewer: ```suggestion
The <xref:Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutCallbackPath%2A> (configuration key: "`SignedOutCallbackPath`") is the request path within the app's base path intercepted by the OpenID Connect handler where the user agent is first returned after signing out from Entra. The sample app doesn't set a value for the path because the default value of "`/signout-callback-oidc`" is used. After intercepting the request, the OpenID Connect handler finally redirects to the <xref:Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutRedirectUri%2A> or <xref:Microsoft.AspNetCore.Authentication.AuthenticationProperties.RedirectUri%2A>, if specified.
```

Your fix to my `SignedOutRedirectUri` link made me realize that the other link I copied from must be outdated, and it is. The last version we shipped `Microsoft.AspNetCore.Builder.OpenIdConnectOptions` was 1.1.

> The <xref:Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutCallbackPath%2A> (configuration key: "`SignedOutCallbackPath`") is the request path within the app's base path intercepted by the OpenID Connect handler where the user agent is first returned after signing out from Entra. The sample app doesn't set a value for the path because the default value of "`/signout-callback-oidc`" is used. After intercepting the request, the OpenID Connect handler finally redirects to the <xref:Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutRedirectUri%2A> or <xref:Microsoft.AspNetCore.Authentication.AuthenticationProperties.RedirectUri%2A>, if specified.

Author: ```suggestion
The <xref:Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutCallbackPath%2A> (configuration key: "`SignedOutCallbackPath`") is the request path within the app's base path intercepted by the OpenID Connect handler where the user agent is first returned after signing out from Entra. The sample app doesn't set a value for the path because the default value of "`/signout-callback-oidc`" is used. After intercepting the request, the OpenID Connect handler redirects to the <xref:Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectOptions.SignedOutRedirectUri%2A> or <xref:Microsoft.AspNetCore.Authentication.AuthenticationProperties.RedirectUri%2A>, if specified.
```

Configure the signed-out callback path in the app's Entra registration. In the Entra or Azure portal, set the path in the **Web** platform configuration's **Redirect URI** entries:

> :::no-loc text="https://localhost/signout-callback-oidc":::

> [!NOTE]
> A port isn't required for `localhost` addresses when using Entra.

If you don't add the signed-out callback path URI to the app's registration in Entra, Entra refuses to redirect the user back to the app and merely asks them to close their browser window.

<!-- UPDATE 10.0 Keep an eye on this NOTE for removal or updates.

>                  The remark on this subject is in the Program files of the
>                  OIDC sample app (Blazor samples repo), and OIDC-BFF sample 
>                  app (Blazor samples repo). -->

Author: ```suggestion
                 The remark on this subject is in the Program file of the
                 Entra sample app (Blazor samples repo). -->
```

> [!NOTE]

> > Entra currently only redirects back to the <xref:Microsoft.AspNetCore.Builder.OpenIdConnectOptions.SignedOutCallbackPath%2A> if the `microsoftonline.com` Authority (`https://login.microsoftonline.com/{TENANT ID}/v2.0/`) is used. This limitation doesn't exist if you can use the "common" Authority with Microsoft Identity Web. For more information, see [postLogoutRedirectUri not working when authority url contains a tenant ID (`AzureAD/microsoft-authentication-library-for-js` #5783)](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/5783).

Reviewer: I know I originally wrote this as part of https://github.com/dotnet/blazor-samples/pull/137/, but I think we can improve it. I later learned that you do not need to use the "common" authority for most users. It's only the tenant admin and external users that will not get redirected back no matter the authority.

```suggestion
> Entra currently does not redirect back to the Blazor application for the primary admin user (root account) and external users. For more information, see [postLogoutRedirectUri not working when authority url contains a tenant ID (`AzureAD/microsoft-authentication-library-for-js` #5783)](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/5783#issuecomment-1465217522).
```

We could also point to this [Q&A answer](https://learn.microsoft.com/en-us/answers/questions/1656146/post-logout-redirect-uri-does-not-work-for-certain#answers:~:text=Microsoft%20Entra%20ID.-,Additionally%2C%20the%20logout%20URL%20is%20not%20currently%20available%20for%20the%20primary%20admin%20user%20(root%20account)%20and%20external%20users.,-Hope%20this%20helps) instead of the GitHub link. I'm not sure which is better. I'm trying to find a more official source.

Author: Should that issue comment (`#issuecomment-1465217522`) be on the link?

https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/5783#issuecomment-1465217522

Reviewer: I'm unsure if that behavior will ever be "fixed". If not, you'd think there'd be a better resource explaining exactly when logout redirects do and don't work that mention how different users are treated differently.

@jennyf19 @jmprieur Do either of you know of a good resource that explains when logout redirects do and do not work and why?

Author: If they don't have a resource to cross-link and it seems like the behavior isn't going change, let's consider dropping the word "currently" and dropping the entire "For more information ... (cross-link)" sentence. It could go something like ...

> [!NOTE]
> Entra doesn't redirect the primary admin user (root account) and external users back to the Blazor application. Instead, Entra recommends that the user close all of their browser windows after logging the user out of the app.

Author: @halter73 ... No answer from your pings. Should we go ahead with either your suggested change or my suggested change? If it's going to be a permanent situation, I think the one I posted without the cross-link is more stable for the article.

Author: I'm going to wrap this up and merge now ... I'm headed out for the holidays ⛄.

I'll use your approach that places the cross-link into the text, but I'll use my wording for the explanation.
